### DISCUSSION TITLE
Using the absolute_url

### DISCUSSION THREAD

**Discussion by edrobinson** [freelawproject/courtlistener#4728] [Q&A]
A While ago i asked about ***HTTP 202 code.***.
Well it's still a PITA.

##### What I am trying to do:
I want to run the search API, then use it's results to find opinions of interest.  

##### About me:  
I use PHP, JS, Smarty Templates and Jaxon PHP package for client <---> server communication.
I am a GitHub and composer fan and, being lazy, use them as much as possible.
I believe in heavily commented code.
I am retired after over 40 years of this "stuff."
I drink too much coffee and work stupid hours.
  
##### Here's what I'm doing:
I have a written a PHP class that runs the API requests. It uses a curl wrapper designed after Python's request and it works well.  
See https://github.com/php-curl-class/php-curl-class .

My class has a separate a method for each AP I want to use and "utility" methods that do all things in support of the API code.  Additional APIs can be added with just a few lines of code as they are all treated similarly.

##### How it works:
A simple web page presents the user with an input for their search query and a button to kick off the search.  
The user types in their query string and clicks the "Go" button.  
I receive the query  and call the search API with it.
The query: "Statutory Co-" , I.C. 72 209(3)  returns 9 hits like this: 
```
      {
            "absolute_url": "\/opinion\/9426580\/smith-v-excel-fabrication-llc\/",
            "attorney": "",
            "caseName": "Smith v. Excel Fabrication, LLC",
            "caseNameFull": "",
            "citation": [],
            "citeCount": 0,
            "cluster_id": 9426580,
            "court": "Idaho Supreme Court",
            "court_citation_string": "Idaho",
            "court_id": "idaho",
            "dateArgued": null,
            "dateFiled": "2023-09-15",
            "dateReargued": null,
            "dateReargumentDenied": null,
            "docketNumber": "48900",
            "docket_id": 67800525,
            "judge": "",
            "lexisCite": "",
            "meta": {
                "timestamp": "2024-06-23T01:13:22.480737Z",
                "date_created": "2023-09-15T16:04:24.963871Z",
                "score": {
                    "bm25": 27.068266
                }
            },
            "neutralCite": "",
            "non_participating_judge_ids": [],
            "opinions": [
                {
                    "author_id": null,
                    "cites": [
                        874422,
                        874758,
                        1124832,
                        1197686,
                        1375098,
                        2510464,
                        2550272,
                        2581698,
                        2598507,
                        3133213,
                        3133236,
                        4198729,
                        5173848,
                        5174602,
                        9795433,
                        9795954
                    ],
                    "download_url": "https:\/\/isc.idaho.gov\/opinions\/48900.pdf",
                    "id": 9840205,
                    "joined_by_ids": [],
                    "local_path": "pdf\/2023\/09\/15\/smith_v._excel_fabrication_llc.pdf",
                    "meta": {
                        "timestamp": "2024-06-25T10:16:03.506659Z",
                        "date_created": "2023-09-15T16:04:24.965026Z"
                    },
                    "per_curiam": false,
                    "sha1": "6941a9af219be737765b89c48e5494a89a255c37",
                    "snippet": "                     IN THE SUPREME COURT OF THE STATE OF IDAHO\n                                    Docket No. 48900\n\n    MITCHELL SMITH, an individual,                         )\n                                                           )\n         Plaintiff-Appellant,                              )\n                                                                  Boise, June 2022 Term\n                                                           )\n    v.                                            ",
                    "type": "combined-opinion"
                }
            ],
            "panel_ids": [],
            "panel_names": [],
            "posture": "",
            "procedural_history": "",
            "scdb_id": "",
            "sibling_ids": [
                9840205
            ],
            "source": "C",
            "status": "Published",
            "suitNature": "",
            "syllabus": ""
        }

 ```
I build an HTML select using the case names and absolute URLs and send it back to the page  
The user clicks an option and I receive the absolute URL.
Now, using that URL, I do a GET on  the CL front end URL like:

https://www.courtlistener.com/opinion/9427848/smith-v-excel-fabrication-llc/  and it returns:  
```
{
    "body": "",
    "raw": "HTTP\/1.1 202 Accepted\r\nServer: CloudFront\r\nDate: Fri, 22 Nov 2024 05:04:42 GMT\r\nContent-Length: 0\r\nConnection: close\r\nx-amzn-waf-action: challenge\r\nCache-Control: no-store, max-age=0\r\nContent-Type: text\/html; charset=UTF-8\r\nAccess-Control-Allow-Origin: *\r\nAccess-Control-Max-Age: 86400\r\nAccess-Control-Allow-Methods: OPTIONS,GET,POST\r\nX-Cache: Error from cloudfront\r\nVia: 1.1 a06140ffee86972bad90c57fc682df36.cloudfront.net (CloudFront)\r\nX-Amz-Cf-Pop: SEA19-C3\r\nAlt-Svc: h3=\":443\"; ma=86400\r\nX-Amz-Cf-Id: Idu9pvW-7fnnP5-UGHaZYRBdgKNg1vMpJGVliTZSbiqLxT-lqvL1vA==\r\n\r\n",
    "headers": {},
    "status_code": 202,
    "protocol_version": 1.1,
    "success": true,
   redirects": 0,
    "url": "https:\/\/www.courtlistener.com\/opinion\/9427848\/smith-v-excel-fabrication-llc\/",
    "history": [],
    "cookies": {}
}
``` 
Click on that url and it displays CL opinion. Drat!  
So, is the 202 code the "I got it and will return something later - or maybe not." code?
If so, how do I deal with it?

I have also tried using the opinion id in the absolute_url to query the opinions API but it doesn't seem to recognize them . . .

Am I even approaching this in a reasonable fashion?  
I still do not have a clear understanding of how the API's relate to each other.
The OPTION requests on them clear up the field uses somewhat but the relations are not clear. 
I need to dig into the data models some more - ya think?  
any help would be appreciated.  
Thanks . . .

**Comment by mlissner** [comment]
Ed, rather than doing a GET on the front end, you should use the API to get the data you want. In this case, you're getting a `cluster_id` back, and so you should get that from the cluster API to get its information. It is also returning nested opinions, which you could get from the opinions API.

**Comment by edrobinson** [comment]
Mike,
I guess I still don't get the absolute_url.  When I get back to that project - after the documentation project - I will use the API.

Thanks

**Reply by mlissner** [reply]
The absolute URL is where it lives on our site as HTML. Presumably, you want the data though, not the full web page.

**Reply by edrobinson** [reply]
Well, that settles that - finally.
